### ISSUE TITLE
xgboost_ray fails on Python 3.8 on macOS

### ISSUE THREAD

**Issue by amogkam** [ray-project/xgboost_ray#41]
In Python 3.8 on macOS, multiprocessing uses a spawn strategy instead of a fork strategy for process creation. This change is no longer compatible with our subclassed `RabitTracker` that uses a process internally rather than a thread and causes `xgboost_ray` training to fail.

Works fine using the `xgboost` built in `RabitTracker`
```
Python 3.8.6 (default, Nov 20 2020, 18:29:40)
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
>>> from ray.services import get_node_ip_address
>>> import xgboost
>>> from xgboost import RabitTracker
>>> node_ip = get_node_ip_address()
>>> rabit_tracker = RabitTracker(hostIP=node_ip, nslave=2)
>>> rabit_tracker.start(nslave=2)
```

Fails using `xgboost_ray` internal `_RabitTracker`
```
Python 3.8.6 (default, Nov 20 2020, 18:29:40)
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
>>> from ray.services import get_node_ip_address
>>> from xgboost_ray.main import _RabitTracker
>>> node_ip = get_node_ip_address()
>>> rabit_tracker = _RabitTracker(hostIP=node_ip, nslave=2)
>>> rabit_tracker.start(nslave=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/amog/dev/xgboost_ray/xgboost_ray/main.py", line 106, in start
    self.thread.start()
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object '_RabitTracker.start.<locals>.run'
```

Once we fix this we should include 3.8 and macOS tests in the CI and make another release.
cc @krfricke

**Comment by krfricke** [comment]
Current master with ray 1.2.0 works fine for me:

```
> python
Python 3.8.6 (default, Apr 14 2021, 14:07:57) 
[Clang 12.0.0 (clang-1200.0.32.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
>>> from ray.services import get_node_ip_address
>>> from xgboost_ray.main import _RabitTracker
>>> node_ip = get_node_ip_address()
>>> rabit_tracker = _RabitTracker(hostIP=node_ip, nslave=2)
>>> rabit_tracker.start(nslave=2)
>>> del rabit_tracker
>>> 
 > python ../../xgboost_ray/examples/simple.py
File descriptor limit 2560 is too low for production servers and may result in connection errors. At least 8192 is recommended. --- Fix with 'ulimit -n 8192'
2021-04-14 14:11:37,731	INFO services.py:1264 -- View the Ray dashboard at http://127.0.0.1:8265
2021-04-14 14:11:39,656	INFO main.py:817 -- [RayXGBoost] Created 4 new actors (4 total actors). Waiting until actors are ready for training.
2021-04-14 14:11:40,582	INFO main.py:860 -- [RayXGBoost] Starting XGBoost training.
(pid=89787) [14:11:40] task [xgboost.ray]:4598953872 got new rank 3
(pid=89777) [14:11:40] task [xgboost.ray]:4556349584 got new rank 2
(pid=89790) [14:11:40] task [xgboost.ray]:4554084240 got new rank 0
(pid=89782) [14:11:40] task [xgboost.ray]:4535186576 got new rank 1
2021-04-14 14:11:41,440	INFO main.py:1304 -- [RayXGBoost] Finished XGBoost training on training data with total N=143 in 1.84 seconds (0.85 pure XGBoost training time).
Final validation error: 0.0210
 > python --version
Python 3.8.6
 > uname -a
Darwin Kais-MacBook-Pro.local 19.6.0 Darwin Kernel Version 19.6.0: Tue Jan 12 22:13:05 PST 2021; root:xnu-6153.141.16~1/RELEASE_X86_64 x86_64
```

@amogkam could you check if it works for you, too, and close the issue if it does?

**Comment by krfricke** [comment]
Ah seems this was fixed in #42 and we only left this issue open. Closing this for now then.
